src/core/attribution/markovChain.test.js

```diff
@@ -143,95 +180,346 @@ describe("core/attribution/markovChain", () => {
       ]);
-      const params: PagerankParams = {chain};
-      const result = await findStationaryDistribution(params, {
-        maxIterations: 255,
-        convergenceThreshold: 1e-7,
-        verbose: false,
-        yieldAfterMs: 1,
-      });
+
+    it("finds a stationary distribution", async () => {
+      const chain = satelliteChain();
+      const params: PagerankParams = {
+        chain,
+        alpha: 0,
+        seed: uniformDistribution(chain.length),
+        pi0: uniformDistribution(chain.length),
+      };
+      const result = await findStationaryDistribution(
+        params,
+        standardOptions()
+      );
 
       expect(result.convergenceDelta).toBeLessThanOrEqual(1e-7);
-      validateConvegenceDelta(chain, result);
+      validateConvergenceDelta(params.chain, params.seed, params.alpha, result);
 
-      expectStationary(chain, result.pi);
+      expectStationary(params.chain, params.seed, params.alpha, result.pi);
       const expected = new Float64Array([1 / 3, 1 / 6, 1 / 6, 1 / 6, 1 / 6]);
       expectAllClose(result.pi, expected);
     });
 
+    it("finds the same stationary distribution regardless of initialDistribution", async () => {
+      const chain = satelliteChain();
+      const alpha = 0.1;
+      const seed = uniformDistribution(chain.length);
+      const initialDistribution1 = singleIndexDistribution(chain.length, 0);
+      const params1 = {chain, alpha, seed, pi0: initialDistribution1};
+      const initialDistribution2 = singleIndexDistribution(chain.length, 1);
+      const params2 = {chain, alpha, seed, pi0: initialDistribution2};
+
+      const result1 = await findStationaryDistribution(
+        params1,
+        standardOptions()
+      );
+
+      const result2 = await findStationaryDistribution(
+        params2,
+        standardOptions()
+      );
+
+      expectAllClose(result1.pi, result2.pi);
+    });
+
+    it("finds a non-degenerate stationary distribution with seed and non-zero alpha", async () => {
+      const chain = satelliteChain();
+      const alpha = 0.1;
+      const seed = singleIndexDistribution(chain.length, 0);
+      const pi0 = uniformDistribution(chain.length);
+      const result = await findStationaryDistribution(
+        {chain, alpha, seed, pi0},
+        standardOptions()
+      );
+
+      expect(result.convergenceDelta).toBeLessThanOrEqual(1e-7);
+      validateConvergenceDelta(chain, seed, alpha, result);
+
+      expectStationary(chain, seed, alpha, result.pi);
+      // determine the correct stationary distribtution via linear algebra
+      // from python3:
+      // >> A = np.array([[0, 0.25, 0.25, 0.25, 0.25],
+      //  [0.5, 0, 0.25, 0, 0.25],
+      //  [0.5, 0.25, 0, 0.25, 0],
+      //  [0.5, 0, 0.25, 0, 0.25],
+      //  [0.5, 0.25, 0, 0.25, 0]])
+      // >> seed = np.array([1, 0, 0, 0, 0])
+      // >> n = len(seed)
+      // >> alpha = .1
+      // >> piStar = alpha * seed * np.linalg.inv(np.eye(n) -(1-alpha)*A)
```

Comment:
I confess that I don’t understand the linear algebra here. I expected to
see either a simple `np.linalg.matrix_power(..., 256)` or a closed-form
solution of the following form:

  - let `B = alpha * seed + (1 - alpha) * A`, so that `B` describes the
    actual Markov process rather than the base chain;
  - find a left eigendecomposition for `B`;
  - let `v` be a nonzero member of the unique eigenspace with
    corresponding eigenvector `1`, guaranteed by the ergodicity theorem;
  - normalize `v` to sum to unity; then
  - `v` is the stationary distribution.

That is, something like:

```python
import numpy as np

# Set up transition matrix
center = [0, 0.25, 0.25, 0.25, 0.25]
even = [0.5, 0, 0.25, 0, 0.25]
odd = [0.5, 0.25, 0, 0.25, 0]
A = np.array([center, even, odd, even, odd])

# Define teleportation parameters
alpha = 0.1
seed = np.array([1, 0, 0, 0, 0])

# Find stationary distribution
B = alpha * seed + (1 - alpha) * A
(w, v) = np.linalg.eig(B.T)
pi = v[:, w.argmax()].T
pi /= pi.sum()
np.testing.assert_allclose(np.matmul(pi, B), pi)

np.set_printoptions(precision=3)
print(repr(pi))  # array([0.379, 0.155, 0.155, 0.155, 0.155])
print(repr(pi * 58))  # array([22.,  9.,  9.,  9.,  9.])
```

But instead I see:

```python
piStar = alpha * seed * np.linalg.inv(np.eye(n) -(1-alpha)*A)
```

Then `piStar` is actually a 5×5 matrix rather than a row vector. Its
first column is indeed the desired stationary distribution, and its
other columns are zero. But if `seed` is set to something else, like
`np.array([0.6, 0.1, 0.1, 0.1, 0.1])`, then the resulting `piStar`
matrix is everywhere nonzero, and none of its rows or columns appears to
be stationary under the Markov chain action. Also, this code blows up if
`alpha = 0.0`, because `eye(5) - A` is singular.

@mzargham, can you help me understand what’s going on here?

Reply:
@wchargin I'd prefer to talk through it on a call, but in a nutshell what I did was solve for the fix point directly using the matrix inverse. There is no iteration; hence no powers of the matrix or eigenvalue decompositions.  I'd love to catch up with you and dig in on some graph theory + dynamical systems anyway.

I wrote the test by generating a case, solving the exact answer in python (with that commented out code) and then just verified that the javascript was converging to that answer.

Reply:
![image](https://user-images.githubusercontent.com/10465438/56506666-11716800-64d4-11e9-9815-0d3520fd47f3.png)

Reply:
Most of the time I would do these derivations with right multiplication and column vectors rather than left multiplication and row vectors so forgive me if its hard to read.

Reply:
Okay, I see now what you mean. The discrepancy is that the included
Python code lists `seed * np.linalg.inv(…)`, which is a Hadamard product
rather than a matrix product, and so the result is ill-shaped (as well
as numerically incorrect). Changing the infix `*` to a call to
`np.linalg.matmul`, or an infix `@` in Python 3, yields the correct
answer.

Thanks for the clarification. Incidentally, this is exactly why I don’t
like having blobs of code in comments: they’re not run or tested, and
thus when they’re wrong they confuse suckers like me who believe that
they’re correct and try to understand them. :-)

> Most of the time I would do these derivations with right
> multiplication and column vectors rather than left multiplication and
> row vectors so forgive me if its hard to read.

No worries; this is perfectly easy to read. :-)

Reply:
@wchargin Makes sense, I'll avoid blobbing code into my code. It was really pseudo scratchwork and I can totally understand why it doesn't belong in the work product even if it was helpful during the work.